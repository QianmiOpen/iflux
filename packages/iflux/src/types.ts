import { EffectLang } from './el';
import { QueryLang } from './ql';
import { RootStore } from './root/store';
import { Store } from './store';

export type TActionHandler = (store: Store<any>, param?: any) => void;
export type TRootActionHandler = (store: RootStore<any>, params?: any) => void;

export type TPath = Array<string | number> | string | QueryLang;

export type TQLang = Array<TPath | Function>;

export type TRelaxPath = Array<
  string | number | Array<string | number> | Object | QueryLang
>;

export interface IQLangProps {
  name: string;
  lang: TQLang;
}

export type TActionRetFn = () => { msg: string; handler: TActionHandler };

export type TAction = (msg: string, handler: TActionHandler) => TActionRetFn;

export interface IStoreProps<T = {}> {
  /**
   * 当前store的命名空间, 如果设置，将会把当前store共享给<Root/>
   */
  ns?: string;

  /**
   * 当前store的debug状态， 默认false
   */
  debug?: boolean;

  /**
   * 当前store的状态，必填
   */
  state?: T;

  /**
   * 当前store的effectlang
   */
  el?: { [name: string]: EffectLang };

  /**
   * 当前store的action
   */
  action?: { [name: string]: TActionRetFn };
}

export interface IRootStoreProps<T = {}> {
  /**
   * 当前store的debug状态， 默认false
   */
  debug?: boolean;

  /**
   * 当前store的状态，必填
   */
  state?: T;

  /**
   * 当前store的effectlang
   */
  el?: { [name: string]: EffectLang };

  /**
   * 当前store的action
   */
  action?: { [name: string]: TActionRetFn };
}

export type TSubscriber = (data: Object) => void;

export interface IProviderProps<T> {
  children?: any;
  store: () => Store<T>;
  onMounted?: (store: Store) => void;
  onWillUnmount?: (store: Store) => void;
  /**
   * 当前应用标记，主要用于调试
   */
  id?: string;
}

export interface IRootProviderProps {
  children?: any;
  store: () => RootStore<any>;
  onMounted?: (store: RootStore) => void;
  onWillUnmount?: (store: RootStore) => void;
}

export type TRelaxProps<T> = T & {
  setState(cb: (state: Object) => void): void;
  dispatch(msg: string, params?: any): void;
};
