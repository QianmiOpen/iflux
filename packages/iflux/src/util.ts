const toString = Object.prototype.toString;

/**
 * 获取当前参数的类型
 * @param param
 */
export const type = (param: any) => toString.call(param);

/**
 * 判断当前的参数是不是字符串
 * @param param
 */
export const isStr = (param: any): param is string =>
  type(param) === '[object String]';

/**
 * 判断当前的参数是不是数组
 * @param param
 */
export const isArray = (param: any): param is Array<any> =>
  type(param) === '[object Array]';

/**
 * 判断参数是不是函数
 */
export const isFn = (param: any): param is Function =>
  type(param) === '[object Function]';

/**
 * 判断参数是不是对象
 */
export const isObj = (param: any): param is object =>
  type(param) === '[object Object]';

/**
 * 根据路径获取对应路径的值
 * 很有immatable获取值得感觉，这样做的好处在于深度获取对象的值
 * 可以不写那么多的值判断，当发现路径中对应的值是undefined时候提前返回
 * 另外，很容易在内部去cache
 *
 * @param data
 * @param path
 */
export const getPathVal = (
  data: Object,
  path: Array<string | number> | string
) => {
  if (isArray(path)) {
    let result = data;
    for (let p of path) {
      result = result[p];
      if (typeof result === 'undefined') {
        return undefined;
      }
    }
    return result;
  } else if (isStr(path)) {
    //support 'list.0.id'
    if (path.indexOf('.') != -1) {
      return getPathVal(data, path.split('.'));
    }
    return data[path];
  }
};
