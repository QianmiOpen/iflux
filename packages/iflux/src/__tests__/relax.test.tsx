import React from 'react';
import render, { act } from 'react-test-renderer';
import { createStore, Provider, Relax, useRelax } from '../index';
import { TRelaxProps } from '../types';
import * as _ from '../util';

type TProps = {
  id: number;
  lname: string;
  name: string;
};

const store = createStore({
  ns: 'relax-test',
  state: {
    name: 'test',
    list: [{ id: 1, name: 'test' }]
  }
});

function TestRelax() {
  const { id, lname, name, setState, dispatch } = useRelax<TProps>([
    ['list', 0, 'id'],
    { lname: 'list.0.name' },
    'name'
  ]);

  return (
    <div>
      {id}
      {name}
      {lname}
      {_.isFn(dispatch) && 'yes'}
      {_.isFn(setState) && 'yes'}
    </div>
  );
}

@Relax([
  //
  ['list', 0, 'id'],
  { lname: 'list.0.name' },
  'name'
])
class RelaxTest extends React.Component {
  relaxProps = {} as TRelaxProps<TProps>;

  render() {
    const { id, name, lname, setState, dispatch } = this.relaxProps;
    return (
      <div>
        {id}
        {name}
        {lname}
        {_.isFn(dispatch) && 'yes'}
        {_.isFn(setState) && 'yes'}
      </div>
    );
  }
}

const TestApp = () => (
  <Provider store={store} debug>
    <TestRelax />
    <RelaxTest />
  </Provider>
);

it('test init', () => {
  const tree = render.create(<TestApp />);
  expect(tree).toMatchSnapshot();

  act(() => {
    global['relax-test'].store.setState(state => {
      state.list[0].id = 2;
    });
  });

  expect(tree).toMatchSnapshot();
});

it('test no subscribe', () => {
  const RelaxApp = () => {
    useRelax([]);
    return <div />;
  };

  const TestApp = () => (
    <Provider store={store}>
      <RelaxApp />
    </Provider>
  );

  const tree = render.create(<TestApp />);
  expect(tree).toMatchSnapshot();
});

it('test defaultInject setState and dispatch', () => {
  const RelaxApp = () => {
    const { setState } = useRelax([]);
    return <div>{_.isFn(setState) && 'yes'}</div>;
  };

  const TestApp = () => (
    <Provider store={store}>
      <RelaxApp />
    </Provider>
  );

  const tree = render.create(<TestApp />).toJSON();
  expect(tree).toMatchSnapshot();
});
