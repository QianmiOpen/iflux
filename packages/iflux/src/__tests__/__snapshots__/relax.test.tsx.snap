// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test defaultInject setState and dispatch 1`] = `
<div>
  yes
</div>
`;

exports[`test init 1`] = `
Array [
  <div>
    1
    test
    test
    test!
    yes
    yes
  </div>,
  <div>
    1
    test
    test
    yes
    yes
  </div>,
]
`;

exports[`test init 2`] = `
Array [
  <div>
    2
    test
    test
    test!
    yes
    yes
  </div>,
  <div>
    2
    test
    test
    yes
    yes
  </div>,
]
`;

exports[`test no subscribe 1`] = `<div />`;
